keps/sig-api-machinery/20190820-master-service-of-type-externalname.md

```diff
@@ -0,0 +1,302 @@
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The master kubernetes service created in the `default` namespace is called `kubernetes`. It's used by Pods running in the cluster to discover the kube-apiserver and communicate with it.
+
+kube-apiserver manages the master `Service` of type `ClusterIP` and associated `Endpoints` objects. Using this service, kubelet injects two environment variables called `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` into Pod's containers. Those environment variables can be detected by Kubernetes API clients such as `client-go` and used to determine if they run as a Pod in the cluster.
+
+This KEP is about adding support in Kubernetes for the master service to be of type `ExternalName` in addition to `ClusterIP`. This allows for kube-apiservers exposed via external FQDN (fully qualified domain name) to be directly consumed by Pods running in the cluster via that FQDN.
+
+## Motivation
```

Comment:
I don't get this. I can understand why you'd want a loadbalancer for external clients, but why make internal clients go through it? What does that gain you? Why not just keep the existing internal apiserver stuff, and add the external loadbalancer as a separate piece?

Reply:
You can't look up the apiserver Endpoints object without a way to contact the apiserver, so there's no workable way to give "internal" clients a special contact path when they've never before contacted the apiserver.

Large clusters need to loadbalance across apiserver instances even during node startup to prevent cascading failure.

Additionally, DNS allows you to rotate apiserver IPs without reinstalling every node.

Reply:
> But cluster IP/kube-proxy provides the load balancing for the internal name. What's not clear is why the clients running in the cluster need to go out to the loadbalancer and hairpin back into the cluster.

Unfortunately, at the moment you can specify only one advertised IP address when starting the `kube-apiserver` (see `--advertise-address` flag in [kube-apiserver](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/)).
So, if your three apiservers are running on `10.0.0.1`, `10.0.0.2`, `10.0.0.3` you are left with several options:

1. run you own advertiser controller which puts in the `default/kubernetes` `Endpoints` those three IP addresses from above.
2. have a loadbalancer (you can use internal with private IP address only) and use it's IP address to configure `--advertise-address`.
3. use DNS-loadbalancing, which has several drawbacks.

For `kube-proxy` to work, it needs to talk to the API server, so you have to chose one of the three options from above.

And there is the problem with `kubelet` - in this case, you have to choose only `2` or `3`, because in lots of Kubernetes setups `kube-proxy` runs as `DaemonSet` and it's started only once `kubelet` is running.

The loadbalancer usage mentioned in this motivation paragraph was only added as an example for common HA setups. While recommended, _not_ all kubernetes clusters today run in HA mode with multiple `kube-apiservers`.

I'll change 

> This allows for kube-apiservers exposed via external FQDN (fully qualified ....

to 

> This allows for kube-apiservers exposed via FQDN (fully qualified ....

Reply:
Ok. So, it wasn't clear to me from the motivation that this really only applies to `kubelet` and `kube-proxy`. Any other executable can use the cluster IP.

So, if that is truly the scope of the problem, perhaps there is a way to fix this that does not affect every Pod. For example, perhaps we can just pass the FQDN to those?

Reply:
The code/functionality modifications should only apply to `kubelet` and `kube-apiserver`.

> For example, perhaps we can just pass the FQDN to those?

Yes, this is a requirement.

Reply:
The change affects every single pod running because every pod gets those env vars. My question was whether we can create a very limited scope change that only affects those two binaries. I am a bit out on a limb here because I haven't gone and looked at the code and at how the bootstrapping of API server access happens in those two executables. But if they are, for example, passed the API server as a flag, then just allowing that flag to be a hostname should be sufficient.

The bigger point is that the wider the scope of affected workloads, the riskier the change. I think that's why this is generating some handwringing - although the change is small in and of itself, it has a wide impact and so a real possibility of unintended consequences.

Reply:
> My question was whether we can create a very limited scope change that only affects those two binaries.

There is nothing to change with regards to bootstrapping for those components.

I've just created a local (`v1.17.0`) cluster with `minikube` - it uses `kubeadm` as a bootstrapper, but any other bootstrapper does the same.

Bellow is the generated `kubeconfig` for `kube-proxy`. It specifies the address of the master node - in this case it's a single node cluster, so a host - `localhost` is used:

```yaml
apiVersion: v1
kind: Config
clusters:
- cluster:
    certificate-authority: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    server: https://localhost:8443
  name: default
contexts:
- context:
    cluster: default
    namespace: default
    user: default
  name: default
current-context: default
users:
- name: default
  user:
    tokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
```

And this is what it kubelet uses after automatic bootstrapping:

```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FUR....
    server: https://localhost:8443
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: system:node:minikube
  name: system:node:minikube@kubernetes
current-context: system:node:minikube@kubernetes
kind: Config
preferences: {}
users:
- name: system:node:minikube
  user:
    client-certificate: /var/lib/kubelet/pki/kubelet-client-current.pem
    client-key: /var/lib/kubelet/pki/kubelet-client-current.pem
```

As you can see those components **already** use a specific kubeconfigs.

Reply:
I am more confused now than ever. If they already use a kubeconfig which already allows server as a URL then I am not following the need for the change at all. They are not even using the env vars.

Reply:
Yes, those 2 components are not using env vars. 

However there are countless other components which use automatic in-cluster apiserver discovery with those environment variables.

Even if you manually create the master service to be of type `ExternalName`, kubelet won't inject those environment variables. See [this issue](https://github.com/kubernetes/kubernetes/issues/79280) for more details.

Reply:
Ok. My early point narrowed the scope of discussion to just those two binaries. That is, I am failing to see the use cases for which this change is *necessary*. Please correct me if I am missing something. Here's my overall logic:

1) Clearly, this change is unnecessary for anything that is not launched by `kubelet` (i.e., runs "out-of-cluster"). In fact this applies to `kubelet` itself.
2) Anything that is launched by `kubelet`, but after `kube-proxy` is up and running and has plumbed the clusterIP can just use the clusterIP, which can appear in the environment variable, and provides load balancing. Thus, this change is unnecessary for all of those workloads.
3) This leaves only `kube-proxy`, and in fact only in the case that it is launched by `kubelet`. According to the above, `kube-proxy` accepts a `kubeconfig`, which in turn has `server:` that can be specified as a URL which may contain a hostname. Therefore, this change is unnecessary for `kube-proxy`.

So, I am failing to see why we *need* the change, and am loathe to take on a risky change that is not *necessary*.

Reply:
That chain of reasoning feels very fragile, and is easily broken by CNIs. You now have three components that need to receive the information in a special way.

I think a large part of the motivation here is cognitive workload of the cluster admin. It's hurting my brain right now trying to figure out what can use the ClusterIP and what can't, what happens if you deallocate the apiserver ClusterIP under each scheme...

Saying "everything asks DNS and a human manually maintains this one DNS record" is appealing. Even if it is publicly resolvable to an internal IP, it's a lot easier to think about.

Reply:
@mvladev and I had a chat over slack and I will take a closer look at how the advertised IP is used versus how endpoints are reconciled, etc.

Reply:
> You can't look up the apiserver Endpoints object without a way to contact the
> apiserver

TTBOMK there are exactly 2 such cases that we have hostorically cared about:
kubelet and kube-proxy.

> DNS allows you to rotate apiserver IPs without reinstalling every node

This is an interesting case, but doesn't seem to warrant changing the access
mode for pods, just the specially bootstrapped components?

> For kube-proxy to work, it needs to talk to the API server, so you have to
> chose one of the three options from above.

There was some discussion about making kubelet and kube-proxy take multiple
values for masters, but I forget how it was resolved.  How it would balance
those would likely be random.

TL;DR there are 2 "special" components which can not use the ClusterIP VIP for
the master - kubelet and kube-proxy.  How to configure those more or less
expects an IP, not a host.  Options:
1) Pick one master IP
2) Set up an IP-exposed load-balancer
3) Set up a hostname-exposed load-balancer

What I don't understand is why that has to be in the `KUBERNETES_SERVICE_HOST`
variable?